ylva.1
.\" Manpage for Ylva.
.\" Any errors or typos, contact niko@byteptr.com.

.TH man 1 "17 Apr 2020" "1.7" "ylva man page"
.SH NAME
ylva \- command line password manager
.SH SYNOPSIS
ylva [FLAGS] <OPTIONS>
.SH DESCRIPTION
Ylva is a command line
password manager program
for Unix-like operating systems.
.SH OPTIONS
.IP "-i, --init <path>"
Initialize new database
.IP "-E, --encrypt"
Encrypt current database
.IP "-D, --decrypt <path>"
Decrypt database
.IP "-c, --copy <id>"
Copy an entry to a new entry
.IP "-a, --add"
Add new entry
.IP "-p, --show-db-path"
Show current database path
.IP "-u, --use-db <path>"
Switch using another database
.IP "-r, --remove <id>"
Remove entry pointed by id
.IP "-f, --find <search>"
Search for entries
.IP "-F, --regex <search>"
Search for entries with regular expressions
.IP "-e, --edit <id>"
Edit entry pointed by id
.IP "-l, --list-entry <id>"
List entry pointed by id
.IP "-t, --show-latest [count]"
Show latest entries, parameter count is optional.
.IP "-A, --list-all"
Show database statuses
.IP "-v, --version"
Show program version
.IP "-g, --gen-password <length>"
Generate password
.IP "-q, --quick <search>"
This is the same as running
--show-passwords -f
.IP "-h, --help"
Show short help and exit
.SH FLAGS
.IP "--auto-encrypt"
Automatically encrypt after exit
.IP "--show-passwords"
Show passwords in listings
.IP "--force"
--force only works with --init option
.SH EXAMPLES
Create a new database:
       ylva --init "/path/to/file.db"
.PP
Open and decrypt database:
       ylva --decrypt "/path/to/existing/file.db"
.PP
Close and encrypt database:
       ylva --encrypt

Ylva knows what database is currently active and encrypts it.
Encrypting will ask you to type a master passphrase which is used for encryption.
.PP
Add an entry to an open database:
       ylva --add
.PP
If you want to export all entries to a file:
       ylva --show-passwords -A > file.txt
.PP
To show latest 10 entries:
       ylva --show-latest 10
.PP
To show all entries ordered by date
       ylva --show-latest
.SH COLORS
Ylva supports colored output. To use colors, set an environment variable
YLVA_COLOR with one of the following value:
BLUE, RED, GREEN, YELLOW, MAGENTA, CYAN or WHITE.
.SH DEFAULT USERNAME
Ylva supports settings default username for entries. To enable it, set
an environment variable YLVA_DEFAULT_USER value to the wanted default.
.SH NOTES
Ylva does not have a concept of "change the master password". When you encrypt
an open database using --encrypt you can type a master password. This password
is then  required to decrypt the database. You can change the master password
every time when you encrypt the database, if you want to.

.SH FILES
.I $HOME/.ylva.lock
.SH AUTHORS
Written by Niko Rosvall.
.SH COPYRIGHT
Copyright (C) 2019-2020 Niko Rosvall <niko@byteptr.com>
.PP
Released under MIT license.


NEWS
Release 1.5 (2019-05-24)
        Password generator now uses RAND_bytes from OpenSSL
        Regex search can now search all the entry fields
        Minor memory leak fixed
        Entries can now be duplicated with -D, --duplicate
        Decrypted database is now only readable by the owner

Release 1.4 (2019-05-18)
        *Titan is now Ylva
        *Ylva no longer supports encrypting individual files or directories, it's not a job for a password manager.
        *Note that if you used colors, change TITAN_COLOR to YLVA_COLOR
        *This release is made mostly because of the name change, but active development will continue
Release 1.3 (2018-09-04)
        *Titan now supports custom count of listing latest entries

Release 1.2 (2017-11-14)
        *Support for colors. See man titan(1) for more information.
        *Added Support encrypting individual files
        *Added support for encrypting directories
        *Many smaller internal changes and bug fixes

ylva.c
/*
 * Copyright (C) 2019-2020 Niko Rosvall <niko@byteptr.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <getopt.h>
#include "cmd_ui.h"
#include "entry.h"
#include "db.h"
#include "utils.h"
#include "pwd-gen.h"
#include "crypto.h"

static int show_password = 0;
static int force = 0;
static int auto_encrypt = 0;

static double v = 1.7;

static void version()
{
    printf("Ylva version %.1f\n", v);
}

static void usage()
{
#define HELP "\
SYNOPSIS\n\
\n\
    ylva [flags] [options]\n\
\n\
OPTIONS\n\
\n\
    -i --init                <path>   Initialize new database\n\
    -E --encrypt                      Encrypt the current password database\n\
    -D --decrypt             <path>   Decrypt password database\n\
    -a --add                          Add new entry\n\
    -c --copy                <id>     Copy an entry\n\
    -r --remove              <id>     Remove entry pointed by id\n\
    -p --show-db-path                 Show current database path\n\
    -u --use-db              <path>   Switch using another database\n\
    -f --find                <search> Search entries\n\
    -F --regex               <search> Search entries with regular expressions\n\
    -e --edit                <id>     Edit entry pointed by id\n\
    -l --list-entry          <id>     List entry pointed by id\n\
    -t --show-latest         <count>  Show latest <count> entries\n\
    -A --list-all                     List all entries\n\
    -h --help                         Show short help and exit. This page\n\
    -g --gen-password        <length> Generate password\n\
    -q --quick               <search> This is the same as running\n\
                                      --show-passwords -f\n\
\n\
    -v --version                      Show version number of program\n\
\n\
FLAGS\n\
\n\
    --auto-encrypt                    Automatically encrypt after exit\n\
    --show-passwords                  Show passwords in listings\n\
    --force                           Ignore everything and force operation\n\
                                      --force only works with --init option\n\
\n\
For more information and examples see man ylva(1).\n\
\n\
AUTHORS\n\
    Copyright (C) 2019-2020 Niko Rosvall <niko@byteptr.com>\n\
"
    printf(HELP);
}

int main(int argc, char *argv[])
{
    int c;

    if(argc == 1)
    {
        usage();
        return 0;
    }

    while(true)
    {
        static struct option long_options[] =
        {
            {"init",                  required_argument, 0,             'i'},
            {"encrypt",               no_argument,       0,             'E'},
            {"decrypt",               required_argument, 0,             'D'},
            {"use-db",                required_argument, 0,             'u'},
            {"add",                   no_argument,       0,             'a'},
            {"copy",                  required_argument, 0,             'c'},
            {"remove",                required_argument, 0,             'r'},
            {"find",                  required_argument, 0,             'f'},
            {"regex",                 required_argument, 0,             'F'},
            {"edit",                  required_argument, 0,             'e'},
            {"list-all",              no_argument,       0,             'A'},
            {"list-entry",            required_argument, 0,             'l'},
            {"gen-password",          required_argument, 0,             'g'},
            {"help",                  no_argument,       0,             'h'},
            {"version",               no_argument,       0,             'v'},
            {"show-db-path",          no_argument,       0,             'p'},
            {"show-latest",           required_argument, 0,             't'},
            {"quick",                 required_argument, 0,             'q'},
            {"auto-encrypt",          no_argument,       &auto_encrypt,  1 },
            {"show-passwords",        no_argument,       &show_password, 1 },
            {"force",                 no_argument,       &force,         1 },
            {0, 0, 0, 0}
        };

        int option_index = 0;

        c = getopt_long(argc, argv, "i:ED:u:ac:r:f:F:e:Al:g:hvpt:q:",
                        long_options, &option_index);

        if(c == -1)
            break;

        switch(c)
        {
        case 0:
            /* Handle flags here automatically */
            break;
        case 'i':
            init_database(optarg, force, auto_encrypt);
            break;
        case 'E': //encrypt
            encrypt_database();
            break;
        case 'D': //decrypt
            decrypt_database(optarg);
            break;
        case 'u':
            set_use_db(optarg);
        case 'a':
            add_new_entry(auto_encrypt);
            break;
        case 'c':
            copy_entry(atoi(optarg));
            break;
        case 'r':
            remove_entry(atoi(optarg), auto_encrypt);
            break;
        case 'f':
            find(optarg, show_password, auto_encrypt);
            break;
        case 'F':
            find_regex(optarg, show_password);
            break;
        case 'e':
            edit_entry(atoi(optarg), auto_encrypt);
            break;
        case 'A':
            list_all(show_password, auto_encrypt, -1);
            break;
        case 'l':
            list_by_id(atoi(optarg), show_password, auto_encrypt);
            break;
        case 'g':
        {
            char *pass = generate_password(atoi(optarg));
            if(pass != NULL)
                free(pass);
            break;
        }
        case 'h':
            usage();
            break;
        case 'v':
            version();
            break;
        case 'p':
            show_current_db_path();
            break;
        case 't':
            show_latest_entries(show_password, auto_encrypt, atoi(optarg));
            break;
        case 'q':
            show_password = 1;
            find(optarg, show_password, auto_encrypt);
            break;
        case '?':
            usage();
            break;
        }
    }
    return 0;
}
